src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -300,6 +284,56 @@ UpdateRelationsToLocalShardTables(Node *node, List *relationShardList)
 								 relationShardList, QTW_EXAMINE_RTES_BEFORE);
 	}
 
+	if (IsA(node, OnConflictExpr))
+	{
+		OnConflictExpr *onConflict = (OnConflictExpr *) node;
+		Oid constraintId = onConflict->constraint;
+
+		if (constraintId != InvalidOid)
+		{
+			Oid constraintRelationId = InvalidOid;
+
+			HeapTuple tp = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraintId));
+			if (HeapTupleIsValid(tp))
+			{
+				Form_pg_constraint contup = (Form_pg_constraint) GETSTRUCT(tp);
+
+				constraintRelationId = contup->conrelid;
+				ReleaseSysCache(tp);
+			}
+
+			/*
+			 * we can return here without calling the walker function, since we know there will be no
+			 * possible tables or constraints after this point, by the syntax.
+			 */
+			if (constraintRelationId == InvalidOid)
+			{
+				return false;
+			}
+
+			RelationShard *relationShard = FindRelationShardByOid(constraintRelationId,
+																  relationShardList);
+
+			if (relationShard != NULL)
```

Comment:
again, Is it possible that relationShard becomes NULL other than buggy cases, should we throw an error for else ?

Reply:
I think for these places an error code such as INVALID_STATE would make sense in the errors. Though I guess postgres doesn't have that defined in its existing error codes (after a really quick look)